### PR TITLE
Fix empty projections panic

### LINKED ISSUE #2551: Empty projection document causes nil pointer deference
### Versions

branch.txt: main
commit.txt: cf0328f6dbb715471908563b62f9ab7c50e77a7d
version.txt: v1.0.0-83-gcf0328f6

### What did you do?

This occurred while we were testing nosqlbooster, it issues this command unsure why.

### How can we reproduce it?

```sh
mongosh --eval 'db.foo.find({}, {});'
```

### What did you expect to see?

No crash

### What did you see instead?

```
2023-05-02T14:36:52.697+0200	DPANIC	// 127.0.0.1:54202 -> 127.0.0.1:27017 	clientconn/conn.go:163	runtime error: invalid memory address or nil pointer dereference
(err = <nil>)
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run.func3
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:163
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:890
runtime.panicmem
	/usr/local/go/src/runtime/panic.go:260
runtime.sigpanic
	/usr/local/go/src/runtime/signal_unix.go:837
github.com/FerretDB/FerretDB/internal/handlers/common.ValidateProjection
	/home/byron/workdir/FerretDB/internal/handlers/common/projection.go:128
github.com/FerretDB/FerretDB/internal/handlers/common.ProjectionIterator
	/home/byron/workdir/FerretDB/internal/handlers/common/projection_iterator.go:32
github.com/FerretDB/FerretDB/internal/handlers/pg.(*Handler).MsgFind.func1
	/home/byron/workdir/FerretDB/internal/handlers/pg/msg_find.go:120
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*Pool).InTransaction
	/home/byron/workdir/FerretDB/internal/handlers/pg/pgdb/transactions.go:90
github.com/FerretDB/FerretDB/internal/handlers/pg.(*Handler).MsgFind
	/home/byron/workdir/FerretDB/internal/handlers/pg/msg_find.go:82
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).handleOpMsg
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:521
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).route
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:404
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:293
github.com/FerretDB/FerretDB/internal/clientconn.acceptLoop.func1
	/home/byron/workdir/FerretDB/internal/clientconn/listener.go:309
2023-05-02T14:36:52.697+0200	WARN	// 127.0.0.1:54202 -> 127.0.0.1:27017 	clientconn/conn.go:155	Failed to set deadline: set tcp 127.0.0.1:27017: use of closed network connection
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run.func2
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:155
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
(err = <nil>)
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc9bbde]

goroutine 88 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x2, 0xc00069e5b0, {0x1c95bf1?, 0x42?, 0xc000277800?})
	/home/byron/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:198 +0xa8
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc00069e5b0, {0x0, 0x0, 0x0})
	/home/byron/go/pkg/mod/go.uber.org/zap@v1.24.0/zapcore/entry.go:264 +0x586
go.uber.org/zap.(*SugaredLogger).log(0xc0002280c8, 0x3, {0x196e62b, 0xd}, {0xc0006741d0, 0x2, 0x2}, {0x0, 0x0, 0x0})
	/home/byron/go/pkg/mod/go.uber.org/zap@v1.24.0/sugar.go:295 +0x13b
go.uber.org/zap.(*SugaredLogger).DPanicf(...)
	/home/byron/go/pkg/mod/go.uber.org/zap@v1.24.0/sugar.go:184
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run.func3()
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:163 +0x195
panic({0x17b4dc0, 0x24150e0})
	/usr/local/go/src/runtime/panic.go:890 +0x263
github.com/FerretDB/FerretDB/internal/handlers/common.ValidateProjection(0xc000832300)
	/home/byron/workdir/FerretDB/internal/handlers/common/projection.go:128 +0x197e
github.com/FerretDB/FerretDB/internal/handlers/common.ProjectionIterator({0x1b741a0?, 0xc000212300}, 0xc0002ec080, 0x0?)
	/home/byron/workdir/FerretDB/internal/handlers/common/projection_iterator.go:32 +0xb2
github.com/FerretDB/FerretDB/internal/handlers/pg.(*Handler).MsgFind.func1({0x1b82890, 0xc000832558})
	/home/byron/workdir/FerretDB/internal/handlers/pg/msg_find.go:120 +0xdae
github.com/FerretDB/FerretDB/internal/handlers/pg/pgdb.(*Pool).InTransaction(0xc00007ac00, {0x1b7bb18, 0xc00085e0c0}, 0xc00072aca8)
	/home/byron/workdir/FerretDB/internal/handlers/pg/pgdb/transactions.go:90 +0x376
github.com/FerretDB/FerretDB/internal/handlers/pg.(*Handler).MsgFind(0xc0002ee000, {0x1b7bb18, 0xc00085e0c0?}, 0x4ae67b?)
	/home/byron/workdir/FerretDB/internal/handlers/pg/msg_find.go:82 +0x96e
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).handleOpMsg(0xc000538360, {0x1b7bb18, 0xc00085e0c0}, 0xbda9b3?, {0xc00048ccdc, 0x4})
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:521 +0x4be
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).route(0xc000538360, {0x1b7bb18, 0xc00085e0c0}, 0xc00048cb60, {0x1b7cde0?, 0xc000058260?})
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:404 +0x16ff
github.com/FerretDB/FerretDB/internal/clientconn.(*conn).run(0xc000538360, {0x1b7bb18, 0xc0007edaa0})
	/home/byron/workdir/FerretDB/internal/clientconn/conn.go:293 +0x129f
github.com/FerretDB/FerretDB/internal/clientconn.acceptLoop.func1()
	/home/byron/workdir/FerretDB/internal/clientconn/listener.go:309 +0xd05
created by github.com/FerretDB/FerretDB/internal/clientconn.acceptLoop
	/home/byron/workdir/FerretDB/internal/clientconn/listener.go:269 +0x73a
task: Failed to run task "run": exit status 2
task: Failed to run task "run": exit status 2
```

### Environment

```markdown
- OS: Ubuntu
- Architecture: x86_64
- Version: 22.10
- Go: go1.19.3 linux/amd64
- Git: 2.37.2
- Deployment: Docker
- Deployment details: 23.0.3
```

### PR DESCRIPTION
**Pull request by w84thesun**
# Description

Closes #2551.

## Readiness checklist

* [ ] I added/updated unit tests.
* [x] I added/updated integration/compatibility tests.
* [ ] I added/updated comments and checked rendering.
* [ ] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

**Review by AlekSi — APPROVED** [review]
:shipit:

File: integration/query_projection_compat_test.go [hunk 1]
```diff
@@ -55,6 +55,10 @@ func TestQueryProjectionCompat(t *testing.T) {
 	providers := append(shareddata.AllProviders(), topLevelFieldsIntegers)
 
 	testCases := map[string]queryCompatTestCase{
+		"EmptyProjection": {
```
Comment (rumyantseva) [inline integration/query_projection_compat_test.go:58]: Can we also add something like

```
		"NilProjection": {
			filter:     bson.D{},
			projection: nil,
		},
```

or it's not possible for the same test?

File: internal/handlers/common/projection.go [hunk 2]
```diff
@@ -35,7 +35,7 @@ var errProjectionEmpty = errors.New("projection is empty")
 func ValidateProjection(projection *types.Document) (*types.Document, bool, error) {
 	validated := types.MakeDocument(0)
 
-	if projection == nil {
+	if projection.Len() == 0 {
```
Comment (rumyantseva) [inline internal/handlers/common/projection.go:38]: What if projection is not set at all? Shouldn't it be `if projection == nil || projection.Len() == 0`?
Or `projection == nil` is not possible?

**Review by rumyantseva — COMMENTED** [review]
The fix looks good to me, but I have one question.

File: internal/handlers/common/projection.go (hunk 2 above)
Comment (AlekSi) [reply internal/handlers/common/projection.go:38]: https://github.com/FerretDB/FerretDB/blob/a1aee0233ffbf7ea170f760d9189618a02e145c8/internal/types/document.go#L131-L135

File: integration/query_projection_compat_test.go (hunk 1 above)
Comment (w84thesun) [reply integration/query_projection_compat_test.go:58]: Thanks for the suggestion, added this test case.

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by AlekSi — APPROVED** [review]

**Pull request merged**